File: Valid_Parentheses.py
# Link: https://leetcode.com/problems/valid-parentheses/

class Solution(object):
    def isValid(self, s):
        stack = []
        dict = {  ')' : '(',  '}' : '{',  ']' : '[' }
        for c in s:
            if c in dict:
                if stack and stack[-1] == dict[c]:
                    stack.pop()
                else: 
                    return False
            else:
                stack.append(c)
        
        return True if not stack else False

File: valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char>st;
        unordered_map<char, char>mp;
        mp[')'] = '(';
        mp['}'] = '{';
        mp[']'] = '[';

        for(auto& bracket: s){
            if(!mp.count(bracket)){
                st.push(bracket);
            }
            else if(!st.empty() && st.top() == mp[bracket]){
                st.pop();
            }
            else{
                return false;
            }
        }
        return st.empty();
    }
};